wfe2/verify.go

```diff
@@ -18,233 +20,511 @@ import (
+	// Convert the account ID string to an int64 for use with the SA's
+	// GetRegistration RPC
+	accountID, err := strconv.ParseInt(accountIDStr, 10, 64)
+	if err != nil {
+		wfe.stats.Inc("Errors.InvalidKeyID", 1)
+		logEvent.AddError("JWS key ID was invalid int64")
+		return nil, nil, probs.Malformed(fmt.Sprintf("Malformed account ID in KeyID header"))
+	}
+
+	// Try to find the account for this account ID
+	account, err := wfe.SA.GetRegistration(ctx, accountID)
+	// If there was an error and it isn't a "Not Found" error, return
+	// a ServerInternal problem since this is unexpected.
+	if err != nil && !berrors.Is(err, berrors.NotFound) {
+		wfe.stats.Inc("Errors.UnableToGetAccountByID", 1)
+		logEvent.AddError(fmt.Sprintf("Error calling SA.GetRegistration: %s", err.Error()))
+		return nil, nil, probs.ServerInternal(fmt.Sprintf(
+			"Error retreiving account %q", accountURL))
+	} else if berrors.Is(err, berrors.NotFound) {
+		// If the account isn't found, return a suitable problem
+		wfe.stats.Inc("Errors.KeyIDNotFound", 1)
+		logEvent.AddError(fmt.Sprintf("Account %q not found", accountURL))
+		return nil, nil, probs.AccountDoesNotExist(fmt.Sprintf(
+			"Account %q not found", accountURL))
+	}
+
+	// Verify the account is not deactivated
+	if features.Enabled(features.AllowAccountDeactivation) && account.Status != core.StatusValid {
+		wfe.stats.Inc("Errors.AccountIsNotValid", 1)
+		logEvent.AddError(fmt.Sprintf("Account %q has status %q", accountURL, account.Status))
+		return nil, nil, probs.Unauthorized(
+			fmt.Sprintf("Account is not valid, has status %q", account.Status))
 	}
 
-	if statName, err := checkAlgorithm(key, parsedJws); err != nil {
+	// Update the logEvent with the account information and return the JWK
+	logEvent.Requester = account.ID
+	logEvent.Contacts = account.Contact
+	return account.Key, &account, nil
+}
+
+// validJWSForKey checks a provided JWS for a given HTTP request validates
+// correctly using the provided JWK. If the JWS verifies the protected payload
+// is returned. The key/JWS algorithms are verified and
+// the JWK is checked against the keyPolicy before any signature validation is
+// done. If the JWS signature validates correctly then the JWS nonce value
+// and the JWS URL are verified to ensure that they are correct.
+func (wfe *WebFrontEndImpl) validJWSForKey(
+	jws *jose.JSONWebSignature,
+	jwk *jose.JSONWebKey,
+	body string,
+	request *http.Request,
+	logEvent *requestEvent) ([]byte, *probs.ProblemDetails) {
+
+	// Check that the public key and JWS algorithms match expected
+	if statName, err := checkAlgorithm(jwk, jws); err != nil {
 		wfe.stats.Inc(statName, 1)
-		return nil, nil, reg, probs.Malformed(err.Error())
+		logEvent.AddError("checkAlgorithm failed: %q", err.Error())
+		return nil, probs.Malformed(err.Error())
+	}
+
+	// If the key doesn't meet the GoodKey policy return a problem immediately
+	if err := wfe.keyPolicy.GoodKey(jwk.Key); err != nil {
```

Comment:
You're right. Good catch. It looks like the existing WFE implementation of `verifyPOST` only checks the key against `GoodKey` for the case where it was supplied by the JWS and not from a registration in the DB. 

I think the analogous thing to do here is only use `GoodKey` from `validSelfAuthenticatedPOST` instead of the generic `validJWSForKey` function that is used by both `validSelfAuthenticatedPOST` and `validPOSTForAccount`. This will avoid calling `GoodKey` on keys that are read from the SA and assumed to have been good at the time they were registered.